src/frontend/hooks/useSyncState.ts

```diff
@@ -0,0 +1,96 @@
+import {MapeoProjectApi} from '@mapeo/ipc';
+import React from 'react';
+import {useProject} from './server/projects';
+
+type SyncState = Awaited<ReturnType<MapeoProjectApi['$sync']['getState']>>;
+
+const projectStateMap = new WeakMap<
+  MapeoProjectApi,
+  ReturnType<typeof createSyncState>
+>();
+
+function identity(state: SyncState | undefined) {
+  return state;
+}
+
+/**
+ * Hook to subscribe to the current sync state. Optionally pass a selector to
+ * subscribe to a subset of the state (to avoid unnecessary re-renders)
+ *
+ * Creates a global singleton for each project, to minimize traffic over IPC -
+ * this hook can safely be used in more than one place without attaching
+ * additional listeners across the IPC channel.
+ *
+ * @example
+ * ```
+ * const connectedPeerCount = useSync(state => state.connectedPeers);
+ * ```
+ *
+ * @param selector Select a subset of the state to subscribe to. Defaults to return the entire state.
+ * @returns
+ */
+export function useSyncState<S = SyncState | undefined>(
+  selector: (state: SyncState | undefined) => S = identity as any,
```

Comment:
not fully, but basically it's to appease TS because it's expecting the `identity` function to map properly to the generics that are being used. not sure if there's an elegant/proper way to express but my hunch is that there isn't  (for example, [Zustand does the same thing](https://github.com/pmndrs/zustand/blob/6f61b76fa096179248a91231789fd5bbc51a36e0/src/react.ts#L55) for the same kind of representation of a state selector). basically, not worth trying to get an identity function default to work precisely on the types level

Reply:
sorry i just realized that you may have been referring to the whole implementation, not just the `as any` part.

the use of the generic is to represent the type of the value returned by the provided `selector` param. by default, if you omit the selector, `useSyncState()` will return the entire `state`. providing the `selector` argument can change the return type of `useSyncState()` though. for example:

```ts
const peersCount = useSyncState((state) => state?.connectedPeers)
```

we want TS to understand that `peersCount` is `number | undefined`, as opposed to default `SyncState | undefined`.

Reply:
Thanks. Yeah, your first instinct was right, i was just asking about the `as any` part!